File: okhttp4/src/test/kotlin/integration_tests/SimpleTest.kt
package integration_tests

import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import notion.api.v1.NotionClient
import notion.api.v1.http.OkHttp4Client
import notion.api.v1.model.pages.Page
import notion.api.v1.model.pages.PageParent
import notion.api.v1.model.pages.PageProperty
import org.junit.Test

typealias prop = PageProperty

class SimpleTest {

  @Test
  fun pages() {
    NotionClient(token = System.getenv("NOTION_TOKEN")).use { client ->
      client.httpClient = OkHttp4Client()

      val databases = client.listDatabases()
      assertTrue { databases.results.isNotEmpty() }

      val database =
          databases.results.find { it.title.any { t -> t.plainText.contains("Test Database") } }!!

      // All the options for "Severity" property (select type)
      val severityOptions = database.properties?.get("Severity")?.select?.options
      // All the options for "Tags" property (multi_select type)
      val tagOptions = database.properties?.get("Tags")?.multiSelect?.options
      // The user object for "Assignee" property (people type)
      val assignee = client.listUsers().results[0] // just picking the first user up

      val newPage: Page =
          client.createPage(
              // Use the "Test Database" as this page's parent
              parent = PageParent.database(databaseId = database.id),
              // Set values to the page's properties
              // (these must be pre-defined before this API call)
              properties =
                  mapOf(
                      "Title" to
                          prop(
                              title =
                                  listOf(
                                      PageProperty.RichText(
                                          text =
                                              PageProperty.RichText.Text(content = "Fix a bug")))),
                      "Severity" to prop(select = severityOptions?.find { it.name == "High" }),
                      "Tags" to prop(multiSelect = tagOptions),
                      "Due" to
                          prop(date = PageProperty.Date(start = "2021-05-13", end = "2021-12-31")),
                      "Velocity Points" to prop(number = 3),
                      "Assignee" to prop(people = listOf(assignee)),
                      "Done" to prop(checkbox = true),
                      "Link" to prop(url = "https://www.example.com"),
                      "Contact" to prop(email = "foo@example.com"),
                  ))
      assertNotNull(newPage)

      val patchResult =
          client.updatePageProperties(
              pageId = newPage.id,
              // Update only "Severity" property
              properties =
                  mapOf(
                      "Severity" to prop(select = severityOptions?.find { it.name == "Medium" }),
                  ))
      assertNotEquals(newPage.lastEditedTime, patchResult.lastEditedTime)

      val fetchedPage = client.retrievePage(newPage.id)
      assertEquals(patchResult.lastEditedTime, fetchedPage.lastEditedTime)
    }
  }
}

File: scala3/src/test/scala/integration_tests/SimpleTest.scala
package integration_tests

import notion.api.v1.ScalaNotionClient
import notion.api.v1.http.JavaNetHttpClient
import notion.api.v1.model.common.PropertyType
import notion.api.v1.model.databases.query.filter.PropertyFilter
import notion.api.v1.model.databases.query.filter.condition.TextFilter
import notion.api.v1.model.pages.PageProperty.RichText
import notion.api.v1.model.pages.PageProperty.RichText.Text
import notion.api.v1.model.pages.{PageParent, PageProperty}
import org.junit.Assert.assertNotNull
import org.junit.Test

import scala.jdk.CollectionConverters._

class SimpleTest {

  @Test
  def databases(): Unit = {
    val client = ScalaNotionClient(System.getenv("NOTION_TOKEN"))
    client.httpClient = new JavaNetHttpClient()
    try {
      val databases = client.listDatabases()
      assertNotNull(databases)

      val databaseId = databases.getResults.asScala
        .filter(
          _.getTitle.asScala.exists(_.getPlainText.contains("Test Database"))
        )
        .head
        .getId

      val database = client.retrieveDatabase(databaseId)
      assertNotNull(database)

      val queryResult =
        client.queryDatabase(
          databaseId,
          filter = {
            val f = new PropertyFilter()
            f.setProperty(PropertyType.Title)
            val title = new TextFilter()
            title.setContains("bug")
            f.setTitle(title)
            f
          },
          pageSize = 1
        )
      assertNotNull(queryResult)
    } finally client.close()
  }

  @Test
  def pages(): Unit = {
    val client = ScalaNotionClient(
      token = System.getenv("NOTION_TOKEN"),
      httpClient = new JavaNetHttpClient()
    )
    try {
      val databases = client.listDatabases()
      assertNotNull(databases)

      val databaseId = databases.getResults.asScala
        .filter(
          _.getTitle.asScala.exists(_.getPlainText.contains("Test Database"))
        )
        .head
        .getId

      val page = client.createPage(
        parent = PageParent.database(databaseId),
        properties = Map(
          "Title" -> {
            val p = new PageProperty()
            p.setTitle(List({
              val rt = new RichText()
              rt.setText(new Text("foo", null))
              rt
            }).asJava)
            p
          }
        )
      )
      assertNotNull(page)

      val updatedPage = client.updatePageProperties(
        pageId = page.getId,
        properties = Map(
          "Title" -> {
            val p = new PageProperty()
            p.setTitle(List({
              val rt = new RichText()
              rt.setText(new Text("This is an important task!", null))
              rt
            }).asJava)
            p
          }
        )
      )
      assertNotNull(updatedPage)

      val fetchedPage = client.retrievePage(page.getId)
      assertNotNull(fetchedPage)

    } finally client.close()
  }

}
